lib/checkstl.cpp

```diff
@@ -843,6 +846,105 @@ static bool isInvalidMethod(const Token * tok)
     return false;
 }
 
+struct InvalidContainerAnalyzer {
+    struct Info {
+        struct Reference {
+            const Token* tok;
+            ErrorPath errorPath;
+        };
+        std::unordered_map<int, Reference> expressions;
+        ErrorPath errorPath;
+        void add(const std::vector<Reference>& refs)
+        {
+            for (const Reference r : refs) {
+                add(r);
+            }
+        }
+        void add(const Reference& r)
+        {
+            if (!r.tok)
+                return;
+            expressions.insert(std::make_pair(r.tok->exprId(), r));
+        }
+
+        std::vector<Reference> invalidTokens() const
+        {
+            std::vector<Reference> result;
+            std::transform(expressions.begin(), expressions.end(), std::back_inserter(result), SelectMapValues{});
+            return result;
+        }
+    };
+    std::unordered_map<const Function*, Info> invalidMethods;
+
+    std::vector<Info::Reference> invalidatesContainer(const Token* tok) const
+    {
+        std::vector<Info::Reference> result;
+        if (Token::Match(tok, "%name% (")) {
+            const Function* f = tok->function();
+            if (!f)
+                return result;
+            ErrorPathItem epi = std::make_pair(tok, "Calling function " + tok->str());
+            const bool dependsOnThis = exprDependsOnThis(tok->next());
+            auto it = invalidMethods.find(f);
+            if (it != invalidMethods.end()) {
+                std::vector<Info::Reference> refs = it->second.invalidTokens();
+                std::copy_if(refs.begin(), refs.end(), std::back_inserter(result), [&](const Info::Reference& r) {
```

Comment:
I wonder if it would be a good idea to capture dependsOnThis by value. Does it actually capture everything in the context or are compilers smart enough and captures only dependsOnThis.

Reply:
It only captures the variables used, but compilers can optimize it further and just capture the stack pointer instead of each variable used. Usually, its inlined so there is no copy(or copy to pointer) or function call(see [this example](https://godbolt.org/z/6vqeqj)).

Reply:
It's good practice to limit the amount of variable captured.